source/common/config/api_type_oracle.cc

```diff
@@ -31,5 +31,26 @@ ApiTypeOracle::getEarlierVersionMessageTypeName(const std::string& message_type)
   }
   return absl::nullopt;
 }
+
+const absl::optional<std::string> ApiTypeOracle::getEarlierTypeUrl(const std::string& type_url) {
```

Comment:
Can't you use this? This may have duplication. https://github.com/envoyproxy/envoy/blob/1323ff23a9372cf5492ea567ea02cca54c891496/source/common/config/resource_name.h#L35

Reply:
Thanks! I didn't know there is a url conversion interface out there.

Reply:
I can't use that template because I don't have the typename of the resources.

Reply:
`TypeUtil::typeUrlToDescriptorFullName` is what you're after I think.

Reply:
I have figured out a way to use helper function.